test/test_broker.hpp

```diff
@@ -1513,12 +1582,30 @@ class test_broker {
                 );
             }
             else {
-                offline_messages.emplace_back(
+                MQTT_NS::optional<std::chrono::steady_clock::duration> message_expiry_interval;
+
+                auto v = get_property<MQTT_NS::v5::property::message_expiry_interval>(props);
+                if (v && v.value().val() != 0) {
+                    message_expiry_interval.emplace(std::chrono::seconds(v.value().val()));
+                }
+
+                auto i = offline_messages.emplace(offline_messages.end(),
                     MQTT_NS::force_move(pub_topic),
                     MQTT_NS::force_move(contents),
                     MQTT_NS::force_move(props),
                     pubopts
                 );
+
+                if(message_expiry_interval) {
+                    i->tim_message_expiry = std::make_shared<as::steady_timer>(ioc, message_expiry_interval.value());
+                    i->tim_message_expiry->async_wait(
+                        [this, i = i](MQTT_NS::error_code ec) mutable {
```

Comment:
Please use weak_ptr pattern. This code is dangerous.

See
https://stackoverflow.com/questions/43045192/how-to-avoid-firing-already-destroyed-boostasiodeadline-timer

1. subscriber is disconnected.
2. publisher publish the message.
3. The message is inserted into offline_messages.
4. timer is set.
5. subscriber is connected again.
6. timer is fired, the timer event is removed from boost asio timer mechanism and the fire event *1 inserted into io_context.
7. offline_messages are sent to subscriber. and erased.
8. event *1 is processed. And access invalidated iterator.

Reply:
Really unexpected that this could. You would.expect some strict ordering from asio, but apparently it does not.